### PR TITLE
Support dotnet pack

### PR DESCRIPTION
**Pull request by zijchen**
Run `dotnet pack` to create a Nuget package containing your DACPAC file.
- No nuspec file needed
- No changes to the project file needed (besides updating the SDK version to the release that will contain this PR)
- DACPAC file will be in the `tools` folder inside the package
- Package will be created to `/bin/debug/ProjectName.1.0.0.nupkg`, configure package ID and version with MSBuild properties: https://learn.microsoft.com/nuget/reference/msbuild-targets#pack-target
- Created package will have `DACPAC` as one of the package types

### CODE REVIEW

File: test/Microsoft.Buld.Sql.Tests/DotnetTestBase.cs [hunk 1]
```diff
@@ -56,22 +57,23 @@ public void EnvironmentSetup()
...
+            {
+                TestUtils.CopyDirectoryRecursive(this.TestDataDirectory, this.WorkingDirectory);
+            }
         }
 
         /// <summary>
-        /// Calls dotnet build to build the project.
+        /// Calls dotnet command on the database project.
         /// </summary>
         /// <param name="arguments">Any additional arguments to be passed to 'dotnet build'.</param>
```
Comment (kisantia) [inline test/Microsoft.Buld.Sql.Tests/DotnetTestBase.cs:69]: nit: update doc comments

File: test/Microsoft.Buld.Sql.Tests/PackTests.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+            // Verify success
+            Assert.AreEqual(0, exitCode, "Pack failed with error " + stdError);
+            Assert.AreEqual(string.Empty, stdError);
+            this.VerifyDacPackage();
+            this.VerifyNugetPackage();
+        }
+
+        [Test]
+        [Description("Verifies packing separately after builid.")]
```
Comment (kisantia) [inline test/Microsoft.Buld.Sql.Tests/PackTests.cs:33]: typo - build

**Review by kisantia — APPROVED** [review]

**Review by dzsquared — APPROVED** [review]

File: test/Microsoft.Buld.Sql.Tests/PackTests.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            Assert.IsTrue(files.Any(f => f.Equals($"tools/{DatabaseProjectName}.dacpac", StringComparison.OrdinalIgnoreCase)),
+                $"Expected 'tools/{DatabaseProjectName}.dacpac' to be in the Nuget package.");
+
+            // Verify the DACPAC package type is set
+            var packageTypes = packageReader.GetPackageTypes();
+            Assert.IsTrue(packageTypes.Any(t => t.Name.Equals("DACPAC", StringComparison.OrdinalIgnoreCase)),
+                "Expected 'DACPAC' to be one of the package types.");
+
+            // Run additional test on the package
+            test?.Invoke(packageReader);
```
Comment (chlafreniere) [inline test/Microsoft.Buld.Sql.Tests/PackTests.cs:169]: why do we need `?` here?

Comment (zijchen) [reply test/Microsoft.Buld.Sql.Tests/PackTests.cs:169]: Optional test parameter, default is null

**Comment by czb182** [comment]
Doesn’t it seem weird to package a dapac, which is a zip file, inside a nupkg which is also a zip file ?

**Comment by chadbaldwin** [comment]
> Doesn’t it seem weird to package a dapac, which is a zip file, inside a nupkg which is also a zip file ?

Probably not. The dacpac is a build artifact, and the expected result of `dotnet pack` is to copy those build artifacts into a nupkg.

Otherwise if you wanted to unzip it first...that would probably impact a lot of existing code on both the packaging and referencing sides.

**Comment by zijchen** [comment]
> > Doesn’t it seem weird to package a dapac, which is a zip file, inside a nupkg which is also a zip file ?
> 
> Probably not. The dacpac is a build artifact, and the expected result of `dotnet pack` is to copy those build artifacts into a nupkg.
> 
> Otherwise if you wanted to unzip it first...that would probably impact a lot of existing code on both the packaging and referencing sides.

Yeah while both dacpac and nupkg are zip files, they have pre-defined underlying structures.

The ultimate goal of packing into nupkg is to be able to consume them as package references. We also plan on releasing the system dacpacs in nupkg form.

**Comment by czb182** [comment]
That would be awesome.  Dacpac package references are long overdue.  With this and sdk style sqlproj it looks like sql is at risk of being a first class citizen: ) now you just need to open source everything

File: test/Microsoft.Buld.Sql.Tests/DotnetTestBase.cs [hunk 4]
```diff
@@ -5,14 +5,15 @@
 using System.Diagnostics;
 using System.IO;
 using System.Text;
+using Microsoft.Build.Construction;
```
Comment (Benjin) [inline test/Microsoft.Buld.Sql.Tests/DotnetTestBase.cs:8]: Is this addition necessary?  Maybe I'm missing something you're using from this namespace.

File: test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs [hunk 5]
```diff
@@ -3,11 +3,33 @@
 
 using Microsoft.Build.Construction;
 using Microsoft.Build.Evaluation;
+using System;
```
Comment (Benjin) [inline test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs:6]: nit: ordering

File: test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs [hunk 6]
```diff
@@ -28,7 +51,10 @@ public static void AddItemGroup(string projectFilePath, string itemName, string[
                     ProjectItemGroupElement itemGroup = project.Xml.AddItemGroup();
                     foreach (string filePath in filePaths)
                     {
-                        itemGroup.AddItem(itemName, filePath);
+                        ProjectItemElement item = project.Xml.CreateItemElement(itemName);
```
Comment (Benjin) [inline test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs:54]: `itemGroup.AddItem()` should return the new item, if you wanted to make this a little more concise.

**Review by Benjin — APPROVED** [review]

File: test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs [hunk 7]
```diff
@@ -3,11 +3,33 @@
...
+        /// <summary>
+        /// Adds a PropertyGroup to the project XML and populates it with <paramref name="properties"/>.
+        /// </summary>
+        public static void AddProperties(string projectFilePath, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            using (ProjectCollection projectCollection = GetNewEngine())
+            {
+                Project project = new Project(projectFilePath, null, "Current", projectCollection, ProjectLoadSettings.IgnoreMissingImports);
+
+                ProjectPropertyGroupElement propertyGroup = project.Xml.AddPropertyGroup();
```
[the remark this answers is not in the document]
Comment (zijchen) [reply test/Microsoft.Buld.Sql.Tests/ProjectUtils.cs:22]: I don't think so, didn't see an overload that accepts a dictionary/collection: https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.construction.projectrootelement.addpropertygroup?view=msbuild-17-netcore#microsoft-build-construction-projectrootelement-addpropertygroup

**Review by Benjin — APPROVED** [review]
